### ISSUE TITLE
Presentation client exits immediately

### ISSUE THREAD

**Issue by SuprDewd** [icpctools/icpctools#323]
I apologize for the vague report, as I didn't get a chance to reproduce this issue and I'm typing this up from memory.

I deployed a small installation of icpctools v2.3.447 with the CDS and Presentation admin running. When I tried starting a Presentation client, it would output the typical startup message, but instead of opening the presentation window it just exited immediately with no error message.

I looked at the previous releases of icpctools and found v2.2.429 to be a little suspect. So I tried redeploying everything from v2.2.428 and ran the exact same commands to spawn the Presentation client, and this time it worked with no issues.

I haven't confirmed that the changes made in v2.2.429 are the true cause of this, but a wild guess is that there is some issue with the reworked connection model. Maybe the main thread exits with no work remaining as soon as the connection thread has been spawned, terminating the program?

This was on Ubuntu 20.04.

**Comment by deboer-tim** [comment]
Hi Bjarki. I don't have access to a Ubuntu machine, but I think I may have found the problem and released a fix. Could you please try the latest prerelease (v2.3.467 or higher) from https://tools.icpc.global and let me know if it fixes the problem?

If not, could you please attach the latest file from the /logs folder and confirm that there is no output to the console at all?
Thanks, Tim

**Comment by SuprDewd** [comment]
I just deployed v2.3.471 and tried running the client again:
```bash
bjarki@skull:~/cds/presentations$ ./client.sh https://127.0.0.1:8443 presentation presentat1on --name Scoreboard
--- ICPC Presentation Client (2.3.471) ---
Log: logs/client_21.03.16_13.36.22.log
The URL did not point to a specific contest, but one contest was found.
Auto-connecting to: https://127.0.0.1:8443/api/contests/test
Scoreboard connecting...
bjarki@skull:~/cds/presentations$ echo $?
0
```
The associated log file is as follows:
```
---- ICPC Presentation Client (2.3.471) ----
OS:  Linux (amd64/5.8.0-44-generic)
JRE: Private Build (14.0.2)
Folder: /home/bjarki/cds/presentations
Host: skull / 10.10.0.2
Locale/TZ: en_US / Greenwich Mean Time
Args: https://127.0.0.1:8443 presentation presentat1on --name Scoreboard 
Log started 2021.03.16 13:36.22
 
2021.03.16 13:36.22 I Option found: --name Scoreboard
2021.03.16 13:36.22 The URL did not point to a specific contest, but one contest was found.
2021.03.16 13:36.22 Auto-connecting to: https://127.0.0.1:8443/api/contests/test
2021.03.16 13:36.22 I Contest source verified
2021.03.16 13:36.22 I Download https://127.0.0.1:8443/presentation/ to /tmp/org.icpc.tools.cache.127001/presentation/listing.dir (200) (8.8kb in 32ms)
2021.03.16 13:36.22 I Updates found on CDS: 1
2021.03.16 13:36.22 I Version check: 2.3.471 (local) vs 2.3.471 (remote)
2021.03.16 13:36.22 I Display 1 found, modes:
2021.03.16 13:36.22 I    Config: false, false, true, DirectColorModel
<repeated a bunch of times>
2021.03.16 13:36.22 I    Config: false, false, true, DirectColorModel
2021.03.16 13:36.22 I Display 2 found, modes:
2021.03.16 13:36.22 I    Config: false, false, true, DirectColorModel
<repeated a bunch of times>
2021.03.16 13:36.22 I    Config: false, false, true, DirectColorModel
2021.03.16 13:36.22 I Presentation Client [Scoreboard] connecting to RESTContestSource[presentation|presentat1on@https://127.0.0.1:8443/api/contests/test]
2021.03.16 13:36.22 Scoreboard connecting...
```

I'm not sure if it's related but I also run into issues if I try to use the standalone version:
```
bjarki@skull:~/cds/presentations$ ./standalone.sh http://127.0.0.1:8080 presentation presentat1on -p scoreboard
--- ICPC Standalone Presentations (2.3.471) ---
Log: logs/standalone_21.03.16_13.40.50.log
Invalid contest source
Must provide a contest source
bjarki@skull:~/cds/presentations$ ./standalone.sh http://127.0.0.1:8080/api/contests/test presentation presentat1on -p scoreboard
--- ICPC Standalone Presentations (2.3.471) ---
Log: logs/standalone_21.03.16_13.41.06.log
Invalid contest source
Must provide a contest source
```
And the associated log file:
```
---- ICPC Standalone Presentations (2.3.471) ----
OS:  Linux (amd64/5.8.0-44-generic)
JRE: Private Build (14.0.2)
Folder: /home/bjarki/cds/presentations
Host: skull / 10.10.0.2
Locale/TZ: en_US / Greenwich Mean Time
Args: http://127.0.0.1:8080/api/contests/test presentation presentat1on -p scoreboard 
Log started 2021.03.16 13:41.06
 
2021.03.16 13:41.06 I ->- Loading presentations ->-
2021.03.16 13:41.06 I -<- Done loading presentations -<-
2021.03.16 13:41.06 E Invalid contest source
2021.03.16 13:41.06 E Must provide a contest source
```
Thanks for the help. Again, there's no urgency here; v2.2.429 seems to be working fine (apart from some intermittent presentation admin crashes), so I will stick to that for NWERC.

On a related note, do you have any documentation on how to hack on ICPC Tools? I tried importing the project into Eclipse and IntelliJ, but I was barely able to compile the project, and wouldn't know how to e.g. run the server with breakpoints.

**Comment by nickygerritsen** [comment]
I have it working in IntelliJ (since I prefer it over eclipse). I can try to write something up on how I did this. It was not easy 😅

**Comment by nickygerritsen** [comment]
See #347 for a PR for the docs for IntelliJ IDEA 😄 .

**Comment by deboer-tim** [comment]
The log should contain several lines with available resolutions and refresh rates like this (see PresentationWindowImpl.java:270 if you want to see the code):
```
2021.03.11 13:33.17 I Display 1 found, modes:
2021.03.11 13:33.17 I    3840x2160 @ 60hz in 32bit
2021.03.11 13:33.17 I    3840x2160 @ 30hz in 32bit
2021.03.11 13:33.17 I    3360x1890 @ 60hz in 32bit
2021.03.11 13:33.17 I    3360x1890 @ 30hz in 32bit
...
2021.03.11 13:33.17 I    Config: false, false, true, DirectColorModel
```

The lack of these is pretty bizarre - the JDK can see the displays but can't tell what mode(s) and resolutions are available. Plus, there's normally only one or two config options, not a bunch of them.
Do you have another JRE you can try, or can you disconnect one monitor temporarily? (Windows has some odd dual-display issues, and I found some Ubuntu JRE bugs from many years ago)

For the standalone client, are you connecting to a CDS or another Contest API? (e.g. DOMjudge) I'm guessing it is just having a problem auto-detecting the contest or auto-switching to https. It may work if you put in the specific URL:

`./standalone.sh https://127.0.0.1:8443/api/contests/test presentation presentat1on -p scoreboard`

**Comment by JobDoesburg** [comment]
I just ran into the same problem. Above v2.2.248, the client is unable to connect with the cds for some reason. The problem only exists for client mode, in standalone there is no problem. More interestingly, it only is a problem on ubuntu/kali/debian systems, I also tested on macOS and there is no problem there.

The logs don't show anything interesting about the connection.
```
2021.09.21 11:50.30 I Display 1 found, modes:
2021.09.21 11:50.30 I    1259x766 @ 60hz in -1bit
2021.09.21 11:50.30 I    1024x768 @ 60hz in -1bit
2021.09.21 11:50.30 I    800x600 @ 60hz in -1bit
2021.09.21 11:50.30 I    640x480 @ 60hz in -1bit
2021.09.21 11:50.30 I    Config: false, false, true, DirectColorModel
2021.09.21 11:50.30 I    Config: false, false, true, DirectColorModel
2021.09.21 11:50.30 I Presentation Client [10] connecting to RESTContestSource[presentation|presentat1on@https://localhost:8443/api/contests/test]
2021.09.21 11:50.30 10 connecting...
```

I have tried v2.3.467, v2.3.471, v2.3.491 (that should be stable) and latest v2.4.562 and all have this error

**Comment by deboer-tim** [comment]
Thanks to Job for showing this problem still exists and ICPC 2020 WF for sending a test laptop, I was able to reproduce and fix via PR #449. Please use pre-release build 2.4.568 or higher.

On my machine I've seen an intermittent problem where the full screen display is offset upward by 37 pixels (??), which is not even the top toolbar height (24 pixels)... not sure if this is due to a bad JDK/machine configuration, but if you see this as well please open a new issue.
